### ISSUE TITLE
VAE code - clarifications

### ISSUE THREAD

**Issue by botev** [casperkaae/probabilistic_ladder#1]
Could the authors confirm or state if there is a mistake in any of the following statements about the code used for VAE:

**Statement 1**
The output of the `DecoderSamplerLayer`  is equal to `z_enc` if `drawdecsample=False`, and does not depend on it if `drawdecsample=True`:

```
def get_output_for(self, input, drawdecsample=False, **kwargs):

        z_enc, mu, var = input

        if drawdecsample:
            batch_size, num_latent = mu.shape
            eps = self._srng.normal(
                [batch_size, self.eq_samples, self.iw_samples, num_latent],
                 dtype=theano.config.floatX)

            z = mu.dimshuffle(0,'x','x',1) + \
                 T.sqrt(var).dimshuffle(0,'x','x',1) * eps

            return z.reshape((-1,num_latent))
        else:
            return z_enc
```

**Statement 2**
If **Statement 1** is correct than this means that in the `VAE` code lines 506-550 the final `ldec_h` is dependable only on `l_z[0]` if `drawdecsample=False` and does not depend on any of the `l_z` variables if `drawdecsample=True`:

```
for i in range(0,num_layers-1)[::-1]:
        ldec_z_mu[i] = denselayer(ldec_h, num_units=latent_sizes[i], W=w_init_mu, nonlinearity=lasagne.nonlinearities.identity, name='ENC_Z_MU%i'%i)
        ldec_z_var[i] = denselayer(ldec_h, num_units=latent_sizes[i], W=w_init_var, nonlinearity=lasagne.nonlinearities.softplus, b=b_init_var, name='ENC_Z_LOG_VAR%i'%i)
        ldec_z = DecoderSampleLayer(l_z[i],mu=ldec_z_mu[i],var=ldec_z_var[i])
        ldec_h = mlp(ldec_z, num_units=hidden_sizes[i], W=w_init_mlp, name='ENC_Z%itoZ%i_DENSE'%(i,i-1), nonlinearity=nonlin_enc, num_mlp_layers=num_mlp_layers)
```

**Statement 3**
If **Statement 2** is correct than `l_dec_x_mu` and `l_dec_x_var` depend only on `l_z[0]` and nothing else, when `drawdecsample=False` which is default and what is used during training:

```
# get output needed for evaluating model with noise if present
train_layers = lasagne.layers.get_output(l_z + lenc_z_mu + lenc_z_var + ldec_z_mu[:-1] + ldec_z_var[:-1] + [l_dec_x_mu, l_dec_x_var], {l_in:sym_x, ldec_mu_in:sym_mu, ldec_var_in:sym_var}, deterministic=False)
z_train = train_layers[:num_layers*1]
z_mu_q_train = train_layers[1*num_layers:2*num_layers]
z_var_q_train = train_layers[2*num_layers:3*num_layers]
z_mu_p_train = train_layers[3*num_layers:4*num_layers-1]
z_var_p_train = train_layers[4*num_layers-1:5*num_layers-2]
x_mu_train = train_layers[5*num_layers-2]
x_var_train = train_layers[5*num_layers-1]
```

**Comment by casperkaae** [comment]
> Statement 1
> The output of the DecoderSamplerLayer is equal to z_enc if drawdecsample=False, and does not depend on it if drawdecsample=True:

The decoder sample layer is just a 'dummy layer' for easy sampling from the generative model. 
- drawdecsample=False it just passes the sample from the variational distribution through
- drawdecsample=True samples from the generative distribution and passes the sample on

> Statement 2
> If Statement 1 is correct than this means that in the VAE code lines 506-550 the final ldec_h is 

dependable only on l_z[0] if drawdecsample=False and does not depend on any of the l_z variables if drawdecsample=True:
No. It passes on a sample from the variational model from the corresponding layer. Its right that the generative layers are disconnected in the sense that they only affect the LL through the KL[pIq] term.

> Statement 3
> If Statement 2 is correct than l_dec_x_mu and l_dec_x_var depend only on l_z[0] and nothing else, 

when drawdecsample=False which is default and what is used during training:
Again the effect on the observation model for the L>0 layers is indirect through the KL[p|q] term in the LL. The top most layers produces a (conditional) prior distribution for the lower layers that deviates more and more from a standard normal prior as you move one layer down in the hierarchy.

**Comment by botev** [comment]
Just to make it abs clear is the diagram below correct, for 3 stochastic latent layers. The dashed lines represent sampling, while the solid ones are deterministic transformations (Nets):
![vae](https://cloud.githubusercontent.com/assets/1889878/19025214/4903308c-890d-11e6-9798-d695cd93326c.jpeg)

PS: Correct typos in the diagram, also note that the indexing in the decoder/generative side are such that they match correctly for which variable they are supposed, rather than the indexing as in the code in the repo.

**Comment by casperkaae** [comment]
Yes thats exactly how a VAE works ( except for the index typos in the inference model)

Btw you actually pinpointed one of the weaknesses of VAEs since higher layers are only trained indirectly through the KL[q|p] term i.e they don't contribute to the reconstruction error directly but only tries to match the distributions of the lower layers

**Comment by botev** [comment]
Hmm, so this points out an interesting fact that there are 2 types of VAEs. What you are describing is when the variational model is directly on `Q(z)`. If we look at https://arxiv.org/abs/1401.4082 from the other camp from DM, not Kingma and Welling, specifically if you check Figure 1a, in fact that is still a VAE, however it is much more closely related to a Ladder Network than the VAE described here (and in fact in most papers these days). However, note that in their case the prior `p(eps_i)` is standard normal inrespectable of the depth for the reason that they are modelling the noise variable and just injecting it along the downaward pass. I think it is still possible to make in their case `q(e_{i+1})` conditioned on `e_i` such that similar to here to be able to inject it on the upward pass. I'm wondering whether it is not needed a comparison of the two approaches. Fundamentally, they are modelling the same process, however exact setup and computational objective is different which could lead to better or worse solutions. 

PS: Note that for 1 stochastic layer (which is unfortunately the predominant thing done in the literature since they could not make it work with more) the two things are equivalent to each other. Only when there are more than one they differ.

**Comment by casperkaae** [comment]
Yes I should have said 'the most common way to parameterise a VAE...'

I'm not sure why the Kingma parameterisation is the most widely used, maybe just because the paper is an easier read than the Rezende one.

**Comment by botev** [comment]
Thanks for the discussion. I will go trough the LVAE now see if I get everything (I think I did from the paper, but code sometimes help to understand in practice what you do). In the VAE I thought initially it is suppose to be like Rezende one (I guess I read mainly their paper for implementation purposes) thus was surprised.

**Comment by botev** [comment]
Sope, the LVAE diagram is correct below? (The red lines are for things which are not always present based on a flag, while the green ones are the LadderMerge layer):
![lvae](https://cloud.githubusercontent.com/assets/1889878/19047380/7670df9c-8999-11e6-879a-8f187842b637.jpeg)

**Comment by casperkaae** [comment]
Yes it looks correct

**Comment by botev** [comment]
Could you just confirm that the results from the paper have been achieved using the default parameters? (e.g. like only_mu_up and etc..)
Btw is 1 epoch a full pass trough the dataset or 1 iteration?

**Comment by casperkaae** [comment]
Hi again

The results from the NIPS version of the paper is calculated using a slightly different code. I'll hopefully get it online soon. I don't think the default parameters gives the results from the paper, but you should be able to plug in the hyperparameter values from the paper and get similar results

Casper

**Comment by botev** [comment]
Hi, the think is in the paper I could not find that smaller details of how everything is parameterized. 
So at least to my understanding in the paper you are sharing the parameters of the generative model and the top down information. However, I think according to equation 13, the connections of the down-top pass you don't use the `mu` but rather the `d_n` going up.

**Comment by casperkaae** [comment]
> So at least to my understanding in the paper you are sharing the parameters of the generative model and the top down information. However, I think according to equation 13, the connections of the down-top pass you don't use the mu but rather the d_n going up.

Yes that right? i'm not completely sure what the question is :)

**Comment by botev** [comment]
Ah sorry I wrote it as a statement. So the questions is since you are not using the mu but the d_n, could you confirm that the following resembles the model with 3 latent variables (e.g. 512-256-128 size of hidden layer and 64-32-16 of latent variables), since I do want to implement it:
1. d_1 = mlp(mlp(x)) / 512 units each
2. mu_enc_1, sigma_enc_1 = linear(d_1), softplus(linear(d_1)) / 64 latent
3. d_2 = mlp(mlp(d_1))/ 256 units each
4. mu_enc_2, sigma_enc_2 = linear(d_2), softplus(linear(d_2)) / 32 latent
5. d_3 = mlp(mlp(d_2)) / 128 units each
6. mu_enc_3, sigma_enc_3 = linear(d_3), softplus(linear(d_3)) / 16 latent
This is for the upward pass, now I assume that the generative model is based on two mlp's as well, so:
7. Sample z_3 ~ mu_enc_3, sigma_enc_3
8. dec_3 = mlp(mlp(z_3)) / 128 units each
9. mu_dec_2, sigma_dec_2 =  linear(dec_3), softplus(linear(dec_3)) / 32 latent
10. mu_2, sigma_2 = combine(mu_enc_2, sigma_enc_2, mu_dec_2, sigma_dec_2)
11. Sample z_2 ~ mu_2, sigma_2
12. dec_2 = mlp(mlp(z_2)) / 256 units each
13. mu_dec_1, sigma_dec_1 =  linear(dec_2), softplus(linear(dec_2)) / 64 latent
14. mu_1, sigma_1 = combine(mu_enc_1, sigma_enc_1, mu_dec_1, sigma_dec_1)
15. Sample z_1 ~ mu_1, sigma_1
16. dec_1 = mlp(mlp(z_1)) / 512 units each
17. p_x = sigmoid(Linear(dec_1))

And you take KL(mu_enc_3, mu_sigma_3| Standard Normal), KL(mu_2, sigma_2 | mu_dec_2, sigma_dec_2), KL(mu_1,sigma_1|mu_dec_1, mu_sigma_1)

**Comment by casperkaae** [comment]
Yes that seems correct. It does not make a big difference if you pass d_# or mu_# up.

**Comment by botev** [comment]
Just for the record, do you get the same (approx similar) results with the LVAE here and the thing as in the paper?

**Comment by casperkaae** [comment]
Yes they should be within 0.2 NATS or something in that order.

**Comment by botev** [comment]
Can I just also ask during training how many samples do you use normally without the fine tuning on MNIST? According to the paper: " fine-tuned by continuing training for 2000 epochs while multiplying the learning rate with 0.75 after every 200 epochs and increase the number of Monte Carlo and importance weighted samples to 10 to reduce the variance in the approximation of the expectations in Eq. (4) and improve the inference model, respectively." . So this means that you start with some samples X before fine tuning and increase it up to 10, the question is is X=1?

**Comment by casperkaae** [comment]
Feel free to ask all the questions you like - I'll try to answer the as fast possible :)

We trained the models using 1 MC sample to approximate the integral over q. Here we used no annealing of the learning rate.

For finetuning we used 10 samples and annealed the learning rate by a factor of 0.75 after every 200 epochs. I think you should be able to obtain similar results faster by annealing the learning rate earlier and also increase the number of samples earlier. Haven't tested that though.

**Comment by leo-liuzy** [comment]
Hi I have a question on code where you are trying to generate latent variable z_1, z_2. Let's say, L=2, I want to generate x.
First, do I use inferred z_1 from q(z_1 | z_2, x) or do I use it from p(z_1 | z_2). Looking from code, it seems it's controled by share_params. There two choise:
1. If I set share_params to be true, decoder mu and var will just copy from encoder mu and var. However, that will lead the KL divergence to be 0. Would that be a problem?
2. If I set share_params to be false, there will be a seperate mu and var, z_1 will be from p(z_1 | z_2). However, in the ELBO derivation, the expectation is w.r.t. z_1 from q(z_1 | z_2, x), this makes me wonder what is the use of z_1 from p(z_1 | z_2)
Please point it out if there is any mistunderstanding
